client/client-multi/client-c/src/kaa/platform-impl/posix/key_utils.c

```diff
@@ -0,0 +1,417 @@
+    uint8_t signature[KAA_SIGNATURE_LENGTH];
+    size_t  encrypted_session_key_length;
+    size_t  session_key_length;
+} endpoint_key_utils_t;
+
+/* Endpoint's RSA Keys */
+static endpoint_key_utils_t keys;
+static mbedtls_pk_context pk_context_;
+static mbedtls_pk_context pk_pub_context;
+
+/* TODO(KAA-1089) */
+__attribute__((weak))
+void ext_get_endpoint_public_key(uint8_t **buffer, size_t *buffer_size)
+{
+    if (!buffer || !buffer_size) {
+        return;
+    }
+    static int key_length;
+    static int initialization = false;
+    static uint8_t buff[KAA_RSA_PUBLIC_KEY_LENGTH_MAX];
+    if (!initialization)
+    {
+        key_length = mbedtls_pk_write_pubkey_der(&pk_pub_context, buff, sizeof(buff));
+        if (key_length < 0) {
+            return;
+	}
+        initialization = true;
+    }
+    *buffer = buff;
+    *buffer_size = key_length;
+}
+
+static int rsa_encrypt(mbedtls_pk_context *pk, const uint8_t *input, size_t input_len, uint8_t *output)
+{
+    int ret = 0;
+
+    mbedtls_entropy_context entropy;
+    mbedtls_ctr_drbg_context ctr_drbg;
+    const uint8_t pers[] = "key_gen";
+
+    mbedtls_ctr_drbg_init(&ctr_drbg);
+    mbedtls_entropy_init(&entropy);
+
+    ret = mbedtls_ctr_drbg_seed(&ctr_drbg, mbedtls_entropy_func, &entropy,
+            pers, sizeof(pers) - 1);
+    if (!ret) {
+        ret = mbedtls_rsa_rsaes_pkcs1_v15_encrypt(mbedtls_pk_rsa(*pk), mbedtls_ctr_drbg_random, &ctr_drbg,
+                MBEDTLS_RSA_PUBLIC, input_len, input, output);
+    }
+
+    mbedtls_ctr_drbg_free(&ctr_drbg);
+    mbedtls_entropy_free(&entropy);
+
+    return ret;
+}
+
+static kaa_error_t set_rsa_public_key(mbedtls_pk_context *pk, const uint8_t *key, size_t key_size)
+{
+    if (!pk || !key || !key_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    if (mbedtls_pk_parse_public_key(pk, key, key_size)!= 0) {
+        return KAA_ERR_INVALID_PUB_KEY;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+static kaa_error_t aes_encrypt_decrypt(int mode, const uint8_t *input, size_t input_size,
+        uint8_t *output, const uint8_t *key)
+{
+    if ((input_size % 16) != 0 || !input) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    if (mode != MBEDTLS_AES_ENCRYPT && mode != MBEDTLS_AES_DECRYPT) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    static bool initialized = false;
+    static mbedtls_aes_context aes_ctx;
+
+    if (!initialized) {
+        mbedtls_aes_init(&aes_ctx);
+        initialized = true;
+    }
+
+    if (mode == MBEDTLS_AES_ENCRYPT) {
+        mbedtls_aes_setkey_enc(&aes_ctx, key, KAA_SESSION_KEY_LENGTH * 8);
+        mbedtls_aes_crypt_ecb(&aes_ctx, MBEDTLS_AES_ENCRYPT, input, output);
+    } else if (mode == MBEDTLS_AES_DECRYPT) {
+        mbedtls_aes_setkey_dec(&aes_ctx, key, KAA_SESSION_KEY_LENGTH * 8);
+        mbedtls_aes_crypt_ecb(&aes_ctx, MBEDTLS_AES_DECRYPT, input, output);
+    }
+
+    return KAA_ERR_NONE;
+}
+
+static int init_aes_key(unsigned char *key, size_t bytes)
```

Comment:
This code seems to be duplicated.
I recommend to share some code between `common/common_key_utils.c` and this file.